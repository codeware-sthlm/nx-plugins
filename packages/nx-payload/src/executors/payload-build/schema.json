{
  "version": 2,
  "title": "Payload build",
  "description": "Build payload admin.",
  "type": "object",
  "properties": {},
  "required": []
}
