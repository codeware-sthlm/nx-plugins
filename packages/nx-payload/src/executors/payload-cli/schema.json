{
  "version": 2,
  "title": "Payload cli",
  "description": "Run payload commands.",
  "type": "object",
  "properties": {
    "outputPath": {
      "type": "string",
      "description": "The output path of the generated files.",
      "x-completion-type": "directory"
    }
  },
  "required": []
}
