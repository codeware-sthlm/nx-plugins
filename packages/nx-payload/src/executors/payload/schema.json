{
  "version": 2,
  "title": "Payload cli",
  "description": "Run payload commands.",
  "type": "object",
  "properties": {},
  "required": []
}
