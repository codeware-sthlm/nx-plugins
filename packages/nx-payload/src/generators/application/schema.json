{
  "$schema": "http://json-schema.org/schema",
  "$id": "NxPayloadApp",
  "title": "Generate a Payload admin application running Express",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the application?",
      "pattern": "^[a-zA-Z].*$"
    },
    "directory": {
      "description": "The path to the new application.",
      "type": "string",
      "x-prompt": "Where would you like to create the application files?"
    },
    "database": {
      "description": "Database adapter to setup initially.",
      "type": "string",
      "enum": ["mongodb", "postgres"]
    },
    "unitTestRunner": {
      "type": "string",
      "enum": ["jest", "none"],
      "description": "Test runner to use for unit tests.",
      "default": "jest"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)",
      "alias": "t"
    },
    "skipFormat": {
      "type": "boolean",
      "description": "Skip formatting application files.",
      "default": false
    },
    "skipE2e": {
      "type": "boolean",
      "description": "Skip generating e2e application.",
      "default": false
    }
  },
  "required": ["directory", "name"],
  "examplesFile": "../../../docs/application-examples.md"
}
