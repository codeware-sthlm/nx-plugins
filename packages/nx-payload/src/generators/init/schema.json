{
  "$schema": "http://json-schema.org/schema",
  "$id": "NxPayloadInit",
  "title": "Init Nx Payload Plugin",
  "type": "object",
  "properties": {},
  "required": []
}
