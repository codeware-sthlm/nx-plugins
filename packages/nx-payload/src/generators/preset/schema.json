{
  "$schema": "http://json-schema.org/schema",
  "$id": "NxPayloadPreset",
  "title": "Nx Payload Preset",
  "type": "object",
  "description": "Generate a Payload CMS application in a integrated workspace. Can be used with `create-nx-workspace --preset=@cdwr/nx-payload`.",
  "properties": {
    "appName": {
      "description": "The name of the application.",
      "type": "string",
      "x-prompt": "What name would you like to use for the application?",
      "pattern": "^[a-zA-Z].*$"
    },
    "appDirectory": {
      "description": "The path to the new application (Generate paths and names `as-provided`).",
      "type": "string",
      "alias": "d"
    },
    "unitTestRunner": {
      "type": "string",
      "enum": ["jest", "none"],
      "description": "Test runner to use for unit tests.",
      "default": "jest"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)",
      "alias": "t"
    },
    "skipE2e": {
      "type": "boolean",
      "description": "Skip generating e2e application.",
      "default": false
    }
  },
  "required": ["appName", "appDirectory"]
}
