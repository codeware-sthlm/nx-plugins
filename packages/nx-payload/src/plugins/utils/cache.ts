import { existsSync } from 'fs';
import { join } from 'path/posix';

import {
  type CreateNodesResult,
  type TargetConfiguration,
  readJsonFile,
  writeJsonFile
} from '@nx/devkit';
import { workspaceDataDirectory } from 'nx/src/utils/cache-directory';

export const cachePath = join(workspaceDataDirectory, 'payload.hash');

/** Global calculated target from either cache or run-time */

export const calculatedTargets: Record<
  string,
  Record<string, Record<string, TargetConfiguration>>
> = {};

const readTargetsCache = (
  cachePath: string
): Record<string, Record<string, TargetConfiguration>> => {
  return process.env.NX_CACHE_PROJECT_GRAPH !== 'false' && existsSync(cachePath)
    ? readJsonFile(cachePath)
    : {};
};

export const targetsCache = readTargetsCache(cachePath);

export const writeTargetsToCache = (
  cachePath: string,
  results: Record<string, TargetConfiguration>
) => writeJsonFile(cachePath, results);

//
// Nx 20+ V2 specifications
//

export const calculatedTargetsV2: Record<
  string,
  Record<string, CreateNodesResult['projects']>
> = {};

const readTargetsCacheV2 = (
  cachePath: string
): Record<string, CreateNodesResult['projects']> => {
  return process.env.NX_CACHE_PROJECT_GRAPH !== 'false' && existsSync(cachePath)
    ? readJsonFile(cachePath)
    : {};
};

export const targetsCacheV2 = readTargetsCacheV2(cachePath);

export const writeTargetsToCacheV2 = (
  cachePath: string,
  results: Record<string, CreateNodesResult['projects']>
) => writeJsonFile(cachePath, results);
