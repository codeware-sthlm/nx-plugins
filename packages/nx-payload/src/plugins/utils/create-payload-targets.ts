import type {
  CreateNodesContext,
  ProjectConfiguration,
  TargetConfiguration
} from '@nx/devkit';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import type { RunCommandsOptions } from 'nx/src/executors/run-commands/run-commands.impl';

import type { BuildExecutorSchema } from '../../executors/build/schema';

import type { NormalizedOptions } from './types';

export const createPayloadTargets = async (
  projectRoot: string,
  projectConfig: ProjectConfiguration,
  options: NormalizedOptions,
  context: CreateNodesContext
): Promise<Record<string, TargetConfiguration>> => {
  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<
    string,
    TargetConfiguration<BuildExecutorSchema | Partial<RunCommandsOptions>>
  > = {};

  // Add `build` target
  targets[options.buildTargetName] = {
    executor: '@cdwr/nx-payload:build',
    inputs: [
      'default',
      Object.hasOwn(namedInputs, 'production') ? '^production' : '^default',
      { externalDependencies: ['express', 'payload'] }
    ],
    outputs: ['{options.outputPath}'],
    options: {
      main: '{projectRoot}/src/main.ts',
      tsConfig: '{projectRoot}/tsconfig.app.json',
      outputPath: '{workspaceRoot}/dist/{projectRoot}',
      outputFileName: 'src/main.js'
    },
    dependsOn: [options.payloadBuildTargetName],
    cache: true
  };

  // Add `mongodb` target
  targets[options.mongodbTargetName] = {
    executor: 'nx:run-commands',
    options: {
      command: `docker ps -q -f name=mongodb-${projectConfig.name} | grep . && echo '[Running] mongodb is already started' || docker run --name mongodb-${projectConfig.name} --rm -d -p 27017:27017 mongo`
    },
    cache: false
  };

  // Add `payload-build` target
  targets[options.payloadBuildTargetName] = {
    executor: '@cdwr/nx-payload:payload-build',
    configurations: {
      production: {
        outputPath: '{workspaceRoot}/dist/{projectRoot}'
      }
    },
    cache: true
  };

  // Add `payload-cli` target
  targets[options.payloadCliTargetName] = {
    executor: '@cdwr/nx-payload:payload-cli',
    cache: false
  };

  // Add `postgres` target
  targets[options.postgresTargetName] = {
    executor: 'nx:run-commands',
    options: {
      command: `docker ps -q -f name=postgres-${projectConfig.name} | grep . && echo '[Running] PostgreSQL init process complete' || docker run --name postgres-${projectConfig.name} --rm --env-file ${projectRoot}/.env -p 5432:5432 postgres`,
      readyWhen: 'PostgreSQL init process complete'
    },
    cache: false
  };

  // Add `start` target
  targets[options.startTargetName] = {
    executor: 'nx:run-commands',
    options: {
      command: `docker compose -f ${projectRoot}/docker-compose.yml up -d`
    },
    cache: false
  };

  // Add `stop` target
  targets[options.stopTargetName] = {
    executor: 'nx:run-commands',
    options: {
      command: `docker compose -f ${projectRoot}/docker-compose.yml down`
    },
    cache: false
  };

  // Add `docker-build` target
  targets[options.dockerBuildTargetName] = {
    executor: 'nx:run-commands',
    options: {
      command: `docker build -f ${projectRoot}/Dockerfile -t ${projectConfig.name} .`
    },
    cache: false
  };

  // Add `docker-run` target
  targets[options.dockerRunTargetName] = {
    executor: 'nx:run-commands',
    options: {
      command: `docker run --name ${projectConfig.name} --rm --env-file ${projectRoot}/.env -d -p 3000:3000 ${projectConfig.name}`
    },
    cache: false
  };

  return targets;
};
