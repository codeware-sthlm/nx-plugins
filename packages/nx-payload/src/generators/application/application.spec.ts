import { type Tree, readJson } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { type PackageJson } from 'nx/src/utils/package-json';

import generator from './application';
import { type AppGeneratorSchema } from './schema';

describe('application generator', () => {
  let tree: Tree;
  const options: AppGeneratorSchema = {
    directory: 'apps/test-dir',
    name: 'test-app'
  };

  console.log = jest.fn();
  console.warn = jest.fn();

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should add payload dependency', async () => {
    await generator(tree, options);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(packageJson.dependencies['payload']).toBeDefined();
  });

  it('should add webpack bundler dependency', async () => {
    await generator(tree, options);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(
      packageJson.dependencies['@payloadcms/bundler-webpack']
    ).toBeDefined();
  });

  it('should add mongodb plugin dependency', async () => {
    await generator(tree, options);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(packageJson.dependencies['@payloadcms/db-mongodb']).toBeDefined();
  });

  it('should add postgres plugin dependency', async () => {
    await generator(tree, options);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(packageJson.dependencies['@payloadcms/db-postgres']).toBeDefined();
  });

  it('should add richtext slate editor plugin dependency', async () => {
    await generator(tree, options);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(
      packageJson.dependencies['@payloadcms/richtext-slate']
    ).toBeDefined();
  });

  it('should add dependencies for express', async () => {
    await generator(tree, options);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(packageJson.dependencies['express']).toBeDefined();
    expect(packageJson.dependencies['@nx/express']).toBeUndefined();
    expect(packageJson.devDependencies['@nx/express']).toBeDefined();
  });

  it('should not add dependencies for mongodb or postgres', async () => {
    await generator(tree, options);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(packageJson.dependencies['mongodb']).toBeUndefined();
    expect(packageJson.dependencies['pg']).toBeUndefined();

    expect(packageJson.devDependencies['mongodb']).toBeUndefined();
    expect(packageJson.devDependencies['pg']).toBeUndefined();
  });

  it('should add payload project files', async () => {
    await generator(tree, options);

    expect(
      tree.exists(`${options.directory}/src/payload.config.ts`)
    ).toBeTruthy();

    expect(tree.exists(`${options.directory}/.eslintrc.json`)).toBeTruthy();

    expect(tree.exists(`${options.directory}/tsconfig.app.json`)).toBeTruthy();
    expect(tree.exists(`${options.directory}/tsconfig.json`)).toBeTruthy();
    expect(tree.exists(`${options.directory}/tsconfig.spec.json`)).toBeTruthy();

    expect(tree.exists(`${options.directory}/webpack.config.js`)).toBeTruthy();
  });

  it('should exclude folders with run-time generated files on build', async () => {
    await generator(tree, options);

    const content: {
      extends: string;
      compilerOptions: object;
      exclude: Array<string>;
      include: Array<string>;
    } = JSON.parse(
      tree.read(`${options.directory}/tsconfig.app.json`, 'utf-8')
    );

    expect(
      content.exclude.filter(
        (e) => e === 'src/generated/*.ts' || e === 'src/migrations/*.ts'
      ).length
    ).toBe(2);

    expect(content.extends).toBe('./tsconfig.json');
    expect(Object.keys(content.compilerOptions).length).toBeGreaterThan(0);
    expect(content.include.length).toBeGreaterThan(0);
  });

  it('should add express project files', async () => {
    await generator(tree, options);

    expect(tree.exists(`${options.directory}/webpack.config.js`)).toBeTruthy();
  });

  it('should create payload e2e application', async () => {
    await generator(tree, options);

    expect(tree.exists(`${options.directory}-e2e/project.json`)).toBeTruthy();
  });

  it('should skip payload e2e application', async () => {
    await generator(tree, { ...options, skipE2e: true });

    expect(tree.exists(`${options.directory}-e2e/project.json`)).toBeFalsy();
  });

  it('should add docker files', async () => {
    await generator(tree, options);

    expect(tree.exists(`${options.directory}/docker-compose.yml`)).toBeTruthy();
    expect(tree.exists(`${options.directory}/Dockerfile`)).toBeTruthy();
    expect(
      tree.exists(`${options.directory}/Dockerfile.dockerignore`)
    ).toBeTruthy();
  });

  it('should add dotenv files', async () => {
    await generator(tree, options);

    expect(tree.exists(`${options.directory}/.env`)).toBeTruthy();
    expect(tree.exists(`${options.directory}/.env.serve`)).toBeTruthy();
  });

  it('should add folders for auto generated files', async () => {
    await generator(tree, options);

    expect(
      tree.exists(`${options.directory}/src/generated/.gitkeep`)
    ).toBeTruthy();

    expect(
      tree.exists(`${options.directory}/src/migrations/.gitkeep`)
    ).toBeTruthy();
  });

  it('should setup mongodb in payload config by default', async () => {
    await generator(tree, { ...options, database: undefined });

    const content = tree.read(
      `${options.directory}/src/payload.config.ts`,
      'utf-8'
    );
    expect(content.match(/mongooseAdapter|MONGO_URL/g).length).toBe(3);
  });

  it('should setup mongodb in payload config', async () => {
    await generator(tree, { ...options, database: 'mongodb' });

    const content = tree.read(
      `${options.directory}/src/payload.config.ts`,
      'utf-8'
    );
    expect(content.match(/mongooseAdapter|MONGO_URL/g).length).toBe(3);
    expect(content.match(/postgresAdapter|POSTGRES_URL/g)).toBeNull();
  });

  it('should setup postgres in payload config', async () => {
    await generator(tree, { ...options, database: 'postgres' });

    const content = tree.read(
      `${options.directory}/src/payload.config.ts`,
      'utf-8'
    );
    expect(content.match(/mongooseAdapter|MONGO_URL/g)).toBeNull();
    expect(content.match(/postgresAdapter|POSTGRES_URL/g).length).toBe(3);
  });

  // @see https://github.com/nrwl/nx/blob/master/packages/remix/src/generators/application/application.impl.spec.ts
  it.todo('should test all options');
});
