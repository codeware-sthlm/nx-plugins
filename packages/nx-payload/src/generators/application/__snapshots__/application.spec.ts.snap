// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application generator should create Dockerfile for npm package manager 1`] = `
"
FROM node:20-alpine as base

FROM base as builder

ARG MONGO_URL
ARG NODE_ENV
ARG PAYLOAD_PUBLIC_SERVER_URL
ARG PORT

ENV MONGO_URL=\${MONGO_URL}
ENV NODE_ENV=\${NODE_ENV}
ENV PAYLOAD_PUBLIC_SERVER_URL=\${PAYLOAD_PUBLIC_SERVER_URL}
ENV PORT=\${PORT}

WORKDIR /app

COPY package.json ./

RUN npm install

COPY . .

RUN npx nx build test-app

# We'll create a structure similar to scaffolding a default payload workspace.
# Admin build output should match \`buildPath\` in payload.config.ts.
FROM base as runtime

ENV NODE_ENV production

WORKDIR /app

COPY --from=builder /app/dist/apps/test-dir/package.json ./

RUN npm install --omit=dev

COPY --from=builder /app/dist/apps/test-dir/src   ./dist
COPY --from=builder /app/dist/apps/test-dir/build ./dist/apps/test-dir/build

EXPOSE 3000

CMD ["node", "dist/main.js"]
"
`;
